java/java.file.launcher/src/org/netbeans/modules/java/file/launcher/queries/LauncherSourceLevelQueryImpl.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.java.file.launcher.queries;
+
+import javax.swing.event.ChangeListener;
+import org.netbeans.api.java.platform.JavaPlatformManager;
+import org.netbeans.modules.java.file.launcher.SharedRootData;
+import org.netbeans.spi.java.queries.SourceLevelQueryImplementation2;
+import org.openide.*;
+import org.openide.filesystems.FileObject;
+import org.openide.util.ChangeSupport;
+import org.openide.util.lookup.ServiceProvider;
+
+/**
+ *
+ * @author lahvac
+ */
+@ServiceProvider(service=SourceLevelQueryImplementation2.class, position=9_999)
+public class LauncherSourceLevelQueryImpl implements SourceLevelQueryImplementation2 {
+
+    @Override
+    public Result getSourceLevel(FileObject javaFile) {
+        //for files, DefaultSourceLevelQueryImpl will respond:
+        if (javaFile.isFolder()) {
```

Comment:
This will only be called (I hope!) if the `ProjectSourceLevelQueryImpl2` will not answer. And SharedData should not be present for files under projects. And `FOQ` will do the same traversing up the file path. So, going through the parents does not seem to do any harm.